asmcomp/dataflow.mli

```diff
@@ -0,0 +1,99 @@
+       normal control flow, falling through the successor(s) of [i].
+     - [at_raise] is the abstract state "after" the instruction for
+       exceptional control flow, branching to the nearest exception handler
+       or exiting the function with an unhandled exception.
+
+     The [transfer] function, then, returns the abstract state "before"
+     the instruction.  The dataflow analysis will, then, propagate this
+     state "before" as the state "after" the predecessor instructions.
+
+     For compound instructions like [Iifthenelse], the [at_next] abstract
+     value that is passed to [transfer] is not the abstract state at
+     the end of the compound instruction (e.g. after the "then" and "else"
+     branches have joined), but the join of the abstract states at
+     the beginning of the sub-instructions.  More precisely:
+     - for [Iifthenelse(tst, ifso, ifnot)], it's the join of the
+       abstract states at the beginning of the [ifso] and [ifnot]
+       branches;
+     - for [Iswitch(tbl, cases)], it's the join of the abstract states
+       at the beginning of the [cases] branches;
+     - for [Icatch(recflag, body, handlers)] and [Itrywith(body, handler)],
+       it's the abstract state at the beginning of [body].
+
+     The [transfer] function is called for every sub-instruction of [instr],
+     as many times as needed to reach a fixpoint.  Hence, it can record
+     the results of the analysis at each sub-instruction in a mutable
+     data structure.  For instance, the transfer function for liveness
+     analysis updates the [live] fields of instructions as a side
+     effect.
+
+     The optional [exnhandler] argument deals with exception handlers.
+     This is a function that transforms the abstract state at the
+     beginning of an exception handler into the exceptional abstract
+     state for the instructions within the body of the handler.
+     Typically, for liveness analysis, it takes the registers live at
+     the beginning of the handler and removes the register
+     [Proc.loc_exn_bucket] that carries the exception value.  If not
+     specified, [exnhandler] defaults to the identity function.
+
+     The optional [constant_transfer] argument can speed up the analysis
+     in a few use cases.  This argument is a predicate over Mach operations.
+     It flags cases where the transfer function returns the same result
+     regardless of the [at_next] and [at_raise] results.  If
+     [constant_transfer] returns [true], the [at_next] abstract state
+     is not computed, and the transfer function is called with [D.bot]
+     for the [at_next] argument.
+     A consequence of this optimization is that some reachable parts
+     of the Mach code are not analyzed at all.  For this reason, the
+     second part of the result (the function [int -> D.t] associating
+     an abstract state to each catch handler label) must not be used,
+     as some handlers have not been analyzed.
```

Comment:
Having spoken to @gretay-js about her concerns with `constant_transfer` (which I share), I wonder whether we could review and merge the version of this PR without this optimisation first, which should give equal or better performance on liveness analysis compared to the current compiler.  Then we could implement safe points using the new framework and make further decisions (about this optimisation and/or others) based on that experience.  In general, I find it much easier to optimise existing correct code, rather than combining the original implementation with optimisation!